src/nsrt_learning/nsrt_learning_main.py

```diff
@@ -113,32 +116,59 @@ def _get_successors(
-                                              (pnad.op.delete_effects,
-                                               "delete")]:
-                for effect in effect_set:
+            for effect in pnad.op.add_effects:
+                if len(pnad.op.add_effects) > 1:
+                    # We don't want sidelining to result in a no-op
                     new_pnad = PartialNSRTAndDatastore(
                         pnad.op.effect_to_side_predicate(
-                            effect, option_vars, add_or_delete),
-                        pnad.datastore, pnad.option_spec)
+                            effect, option_vars, "add"), pnad.datastore,
+                        pnad.option_spec)
                     sprime = list(s)
                     sprime[i] = new_pnad
                     yield (None, tuple(sprime), 1.0)
+
             # ...consider removing it.
             sprime = list(s)
             del sprime[i]
             yield (None, tuple(sprime), 1.0)
 
-    score_func = _PredictionErrorScoreFunction(predicates, [], {}, train_tasks)
+    if CFG.side_predicate_learner == "prediction_error_hillclimbing":
+        score_func = _PredictionErrorScoreFunction(predicates, [], {},
+                                                   train_tasks)
+
+        def _evaluate(s: Tuple[PartialNSRTAndDatastore, ...]) -> float:
+            # Score function for search. Lower is better.
+            strips_ops = [pnad.op for pnad in s]
+            option_specs = [pnad.option_spec for pnad in s]
+            score = score_func.evaluate_with_operators(frozenset(),
+                                                       trajectories,
+                                                       segmented_trajs,
+                                                       strips_ops,
+                                                       option_specs)
+            return score
+
+    elif CFG.side_predicate_learner == "preserve_skeletons":
 
-    def _evaluate(s: Tuple[PartialNSRTAndDatastore, ...]) -> float:
-        # Score function for search. Lower is better.
-        strips_ops = [pnad.op for pnad in s]
-        option_specs = [pnad.option_spec for pnad in s]
-        score = score_func.evaluate_with_operators(frozenset(), trajectories,
-                                                   segmented_trajs, strips_ops,
-                                                   option_specs)
-        return score
+        def _evaluate(s: Tuple[PartialNSRTAndDatastore, ...]) -> float:
+            # Score function for search. Lower is better.
+            strips_ops = [pnad.op for pnad in s]
+            option_specs = [pnad.option_spec for pnad in s]
+            preserves_harmlessness = check_harmlessness(
+                predicates, train_tasks, trajectories, segmented_trajs,
+                strips_ops, option_specs)
+            # NOTE: Arbitrary large number bigger than total number of
+            # operators.
```

Comment:
here, i think we should say "bigger than the total number of operators that we had at the start of search" -- is that right?